File: Tyche.iOS/Account/Sources/Account/Login/LoginViewModel.swift
import Foundation
import Core
import UI
import Session

public class LoginViewModel: ObservableObject {
    @Published @MainActor private(set) var state: LodableViewState<AccountBundle> = .initial
    
    private let loginUseCase: LoginUseCase
    
    public init(loginUseCase: LoginUseCase) {
        self.loginUseCase = loginUseCase
    }
    
    func resetState() {
        Task {
            await MainActor.run {
                state = .initial
            }
        }
    }
    
    func login(loginCredential: LoginCredentialModel) {
        Task {
            await MainActor.run {
                state = .loading
            }
            
            let result = await loginUseCase.execute(loginCredential: loginCredential.toLoginCredential())
            switch result {
            case .success(let userProfile):
                await MainActor.run {
                    state = .success(userProfile)
                }
            case .failure(let error):
                await MainActor.run {
                    state = .failure(error.toLocalizedError())
                }
            }
        }
    }
}

private extension Error {
    func toLocalizedError() -> Error {
        if let loginError = self as? LoginError {
            return loginError.toLoginLocalizedError()
        } else if let networkError = self as? NetworkError {
            return networkError.toNetworkLocalizedError()
        }
        
        return UnknownLocalizedError()
    }
}

File: Tyche.Android/account/src/main/java/com/felipearpa/tyche/account/login/LoginViewModel.kt
package com.felipearpa.tyche.account.login

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.felipearpa.tyche.session.AccountBundle
import com.felipearpa.tyche.session.authentication.application.LoginUseCase
import com.felipearpa.tyche.ui.exception.orLocalizedException
import com.felipearpa.tyche.ui.state.LoadableViewState
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class LoginViewModel @Inject constructor(private val loginUseCase: LoginUseCase) : ViewModel() {
    private val _state =
        MutableStateFlow<LoadableViewState<AccountBundle>>(LoadableViewState.Initial)
    val state: Flow<LoadableViewState<AccountBundle>>
        get() = _state.asStateFlow()

    fun reset() {
        _state.value = LoadableViewState.Initial
    }

    fun login(loginCredential: LoginCredentialModel) {
        viewModelScope.launch {
            _state.emit(LoadableViewState.Loading)
            loginUseCase.execute(loginCredential = loginCredential.toLoginCredential())
                .onSuccess { accountBundle -> _state.emit(LoadableViewState.Success(accountBundle)) }
                .onFailure { exception ->
                    _state.emit(
                        LoadableViewState.Failure(
                            exception.toLoginLocalizedException().orLocalizedException()
                        )
                    )
                }
        }
    }
}